packages/authentication/src/providers/username-password.ts

```diff
@@ -13,8 +18,48 @@ class UsernamePasswordProvider extends AbstractAuthenticationModuleProvider {
     this.authUserSerivce_ = AuthUserService
   }
 
-  async authenticate(userData: Record<string, unknown>) {
-    return {}
+  async authenticate(
+    userData: Record<string, any>
+  ): Promise<AuthenticationResponse> {
+    const { email, password } = userData.body
+
+    if (typeof password !== "string") {
+      return {
+        success: false,
+        error: "Password should be a string",
+      }
+    }
+
+    if (typeof email !== "string") {
+      return {
+        success: false,
+        error: "Email should be a string",
+      }
+    }
+
+    const authUser = await this.authUserSerivce_.retrieveByProviderAndEntityId(
+      email,
+      UsernamePasswordProvider.PROVIDER
+    )
+
+    const password_hash = authUser.provider_metadata?.password
+
+    if (password_hash && typeof password_hash === "string") {
+      const buf = Buffer.from(password_hash, "base64")
+
+      const success = await Scrypt.verify(buf, password)
```

Comment:
suggestion: experiment running that with `worker_threads` (it doesn't have to be in this PR)